### ISSUE TITLE
Thenable coercion should use thenable.then(deferred.resolve, deferred.reject), not deferred.resolve(thenable)

### ISSUE THREAD

**Issue by rossberg** [domenic/promises-unwrapping#77]
I cannot make any sense of the change that is supposed to fix issue #62. Especially, in the new step 3.i, you resolve the current promise with the coerced one as its _value_? That doesn't seem to make sense, you have to invoke its chaining method, just like in the case where lookup fails.

Also, you enter _every_ thenable into the coercion map, even ones that are promises already. Is that intentional?

**Comment by domenic** [comment]
Thenables are demonstrably not broken, as all the passing the tests show.

> Especially, in the new step 3.i, you resolve the current promise with the coerced one as its value?

Resolving a promise to another promise essentially makes it behave as that promise, so I don't see the problem. All the tests pass; could you give an example of a test that fails?

> Also, you enter every thenable into the coercion map, even ones that are promises already. Is that intentional?

Hmm, it looks like only one of the calls to UpdateDeferredFromPotentialThenable is guarded by IsPromise beforehand. That is probably a bug.

**Comment by rossberg** [comment]
> Resolving a promise to another promise essentially makes it behave as that promise, so I don't see the problem.

They only behave "as that promise" wrt to consecutive .then, not wrt .flatMap. In other words, this breaks the internal invariants necessary for interop. Consider:

```
  let thenable = {then: (f) => f(5)}
  let {promise, resolve} = Promise.deferred()
  promise.then(() => thenable).flatMap(print)  // (1)
  promise.then(() => thenable).flatMap(print)  // (2)
  resolve(1)
```

As far as I can see, line (1) will result in printing "5" as expected, but the second time round in line (2) the same code would suddenly print "[object Object]" (because what the second print gets is actually the promise created by the first .then).

(I suspect you can also observe differences with .then alone, because the order in which pending handler tasks are queued will differ slightly depending on the thenable caching. But that's more subtle, and I don't have access to my system right now to try constructing an example.)

> Hmm, it looks like only one of the calls to UpdateDeferredFromPotentialThenable is guarded by IsPromise beforehand. That is probably a bug.

Honestly, I don't get why there need to be two such calls in the first place. Isn't the partial duplication of work that is going on between Promise Reaction, Promise Resolution, and UpdateDeferredFromPotentialThenable somewhat odd?

**Comment by domenic** [comment]
> They only behave "as that promise" wrt to consecutive .then, not wrt .flatMap.

Ah, you're right, certainly. I am not used to thinking on the monad level, only the promise level. Thanks for the correction; it should be a pretty straightforward change.

> Honestly, I don't get why there need to be two such calls in the first place. Isn't the partial duplication of work that is going on between Promise Reaction, Promise Resolution, and UpdateDeferredFromPotentialThenable somewhat odd?

I agree it's annoying and have tried to consolidate it over time into UpdateDeferredFromPotentialThenable. But it is necessary, since `then` needs to do unwrapping both on the input side (unwrapping the internal resolution value before passing it to `onFulfilled` or `onRejected`) and on the output side (unwrapping the return value before storing it internally). The former is necessary to protect `then` users from promises-for-promises; the latter is necessary to avoid the excess memory consumption @erights worries about.

**Comment by domenic** [comment]
I have renamed this issue to reflect the main bug at hand, and created a new issue, #79, to track the issues with the thenable coercion weak map. This issue should be a trivial fix, but #79 is harder.
